File: src/main/java/me/knighthat/interactivedeck/component/ibutton/IButton.java
/*
 * Copyright (c) 2023. Knight Hat
 * All rights reserved.
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use,copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT.IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
 * OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package me.knighthat.interactivedeck.component.ibutton;

import com.google.gson.JsonObject;
import lombok.Getter;
import me.knighthat.interactivedeck.menus.MenuProperty;
import me.knighthat.interactivedeck.task.Task;
import me.knighthat.lib.component.LiveComponent;
import me.knighthat.lib.component.ibutton.InteractiveButton;
import me.knighthat.lib.connection.request.RequestJson;
import me.knighthat.lib.connection.request.TargetedRequest;
import me.knighthat.lib.json.JsonSerializable;
import me.knighthat.lib.logging.EventLogging;
import me.knighthat.lib.logging.Log;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.io.IOException;
import java.util.Objects;
import java.util.UUID;

import static me.knighthat.interactivedeck.utils.ColorUtils.TRANSPARENT;

@Getter
public class IButton extends JComponent implements InteractiveButton, JsonSerializable, RequestJson {

    /**
     * Turns {@link JsonObject} to {@link IButton} instance.<br>
     *
     * @param profile this button belongs to.
     * @param json    the json text of this button.
     *
     * @return instance of {@link IButton}.
     *
     * @throws IOException if coordinate <b>x</b> or <b>y</b> is missing.
     */
    public static @NotNull IButton fromJson( @NotNull UUID profile, @NotNull JsonObject json ) throws IOException {
        if (!json.has( "x" ))
            throw new IOException( "Missing coordinate \"x\"" );
        if (!json.has( "y" ))
            throw new IOException( "Missing coordinate \"y\"" );

        int x = json.get( "x" ).getAsInt();
        int y = json.get( "y" ).getAsInt();
        UUID uuid;
        if (json.has( "uuid" )) {
            String idString = json.get( "uuid" ).getAsString();
            uuid = UUID.fromString( idString );
        } else {
            Log.warn( "Button from profile " + profile + " does not have an UUID, assigning new one." );
            uuid = UUID.randomUUID();
        }

        IButton button = new IButton( uuid, profile, x, y, null );
        button.update( json );

        return button;
    }

    private final @NotNull UUID uuid;
    private final @NotNull UUID profile;
    private final int posX;
    private final int posY;
    private final @NotNull IButtonBackground back;
    private final @NotNull IButtonForeground front;
    private @Nullable Task task;

    IButton( @NotNull UUID uuid,
             @NotNull UUID profile,
             int posX, int posY,
             @Nullable Task task ) {
        this.uuid = uuid;
        this.profile = profile;
        this.back = new IButtonBackground( this );
        this.front = new IButtonForeground( this );
        this.posX = posX;
        this.posY = posY;
        this.task = task;

        setForeground( TRANSPARENT );

        setLayout( new OverlayLayout( this ) );
        add( front, 0 );
        add( back, 1 );
    }

    public IButton( @NotNull UUID profile, int posX, int posY ) {
        this( UUID.randomUUID(), profile, posX, posY, null );
    }

    public void setTask( @Nullable Task task ) {
        if (Objects.equals( this.task, task ))
            return;
        logAndSendUpdate( "task", this.task, task );
        this.task = task;
    }

    @Override
    public void logAndSendUpdate( @NotNull String property, @Nullable Object oldValue, @Nullable Object newValue ) {EventLogging.DefaultImpls.logAndSendUpdate( this, property, oldValue, newValue );}

    @Override
    public void remove() {MenuProperty.remove( this );}

    @NotNull
    @Override
    public TargetedRequest.Target getTarget() {return TargetedRequest.Target.BUTTON;}

    @Override
    public void logUpdate( @NotNull String property, @Nullable Object oldValue, @Nullable Object newValue ) {Log.buttonUpdate( uuid, property, oldValue, newValue );}

    @Override
    public void sendUpdate( @NotNull String property, @Nullable Object oldValue, @Nullable Object newValue ) {LiveComponent.DefaultImpls.sendUpdate( this, property, oldValue, newValue );}

    @Override
    public void update( @NotNull JsonObject json ) {
        if (json.has( "icon" ))
            back.update( json.getAsJsonObject( "icon" ) );
        if (json.has( "label" ))
            front.update( json.getAsJsonObject( "label" ) );
        if (json.has( "task" ))
            task = Task.fromJson( json.getAsJsonObject( "task" ) );
    }

    @NotNull
    @Override
    public JsonObject toRequest() {
        JsonObject json = serialize();
        json.add( "icon", back.toRequest() );
        return json;
    }

    @Override
    public @NotNull JsonObject serialize() {
        /* Template
         * {
         *      "uuid": $uuid,
         *      "x": $x,
         *      "y": $y,
         *      "task":
         *      {
         *          Task.json()
         *      }
         *      "icon":
         *      {
         *          BIcon.json()
         *      }
         *      "label":
         *      {
         *          BLabel.json()
         *      }
         * }
         */

        JsonObject json = new JsonObject();

        json.addProperty( "uuid", uuid.toString() );
        json.addProperty( "x", posX );
        json.addProperty( "y", posY );
        if (task != null)
            json.add( "task", task.serialize() );
        json.add( "icon", back.serialize() );
        json.add( "label", front.serialize() );

        return json;
    }
}

File: src/main/java/me/knighthat/interactivedeck/component/ibutton/IButton.kt
/*
 * Copyright (c) 2023. Knight Hat
 * All rights reserved.
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use,copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package me.knighthat.interactivedeck.component.ibutton

import com.google.gson.JsonObject
import me.knighthat.interactivedeck.persistent.Persistent
import me.knighthat.interactivedeck.task.Task
import me.knighthat.interactivedeck.utils.ColorUtils
import me.knighthat.lib.component.ibutton.InteractiveButton
import me.knighthat.lib.connection.request.RequestJson
import me.knighthat.lib.json.JsonSerializable
import me.knighthat.lib.logging.Log
import me.knighthat.lib.task.ClientTask
import me.knighthat.lib.util.ShortUUID
import java.awt.Color
import java.awt.Container
import java.awt.Font
import java.io.IOException
import java.util.*
import javax.swing.OverlayLayout

class IButton(
    override val uuid: UUID,
    override val profile: UUID,
    override val posX: Int,
    override val posY: Int,
) : Container(), InteractiveButton, JsonSerializable, RequestJson {

    companion object {
        /**
         * Turns [JsonObject] to [IButton] instance.
         *
         * @param profile this button belongs to.
         * @param json    the json text of this button.
         *
         * @return instance of [IButton].
         *
         * @throws IOException if coordinate **x** or **y** is missing.
         */
        @Throws(IOException::class)
        fun fromJson(profile: UUID, json: JsonObject): IButton {
            if (!json.has("x"))
                throw IOException("Missing coordinate \"x\"")
            if (!json.has("y"))
                throw IOException("Missing coordinate \"y\"")

            val x = json["x"].asInt
            val y = json["y"].asInt
            val uuid: UUID =
                    if (json.has("uuid")) {
                        val idString = json["uuid"].asString
                        UUID.fromString(idString)
                    } else {
                        Log.warn("Button [x=$x,y=$y] from profile $profile does not have an UUID, assigning new one.")
                        UUID.randomUUID()
                    }

            val button = IButton(uuid, profile, x, y)
            button.update(json)

            return button
        }
    }

    private val back: ButtonBackground = ButtonBackground(uuid)
    private val front: ButtonForeground = ButtonForeground(uuid)

    // Properties
    var border: Color
        get() = back.border
        set(value) {
            back.border = value
        }
    var fill: Color
        get() = back.fill
        set(value) {
            back.fill = value
        }
    var label: String
        get() = front.label
        set(value) {
            front.label = value
        }
    var styler: Font
        get() = front.styler
        set(value) {
            front.styler = value
        }
    var fontColor: Color
        get() = front.fontColor
        set(value) {
            front.fontColor = value
        }
    var task: Task? = null
        set(value) {
            // If the new task is the same as the current one, then do nothing
            if (Objects.equals(task, value)) return

            logAndSendUpdate("task", task, value)
            field = value
        }

    init {
        foreground = ColorUtils.TRANSPARENT

        layout = OverlayLayout(this)
        add(front, 0)
        add(back, 1)

        Log.deb("Created button ${ShortUUID.from(uuid)} at [x=$posX,y=$posY]")
    }

    constructor(profile: UUID, posX: Int, posY: Int) : this(UUID.randomUUID(), profile, posX, posY)

    fun toggleSelect() {
        back.isSelected = !back.isSelected
    }

    override fun update(json: JsonObject) {
        for (entry in json.entrySet()) {

            val value = entry.value
            if (value !is JsonObject)
                continue

            when (entry.key) {
                "icon"  -> back.update(value)
                "label" -> front.update(value)
                "task"  -> task = Task.fromJson(value)
            }
        }
    }

    override fun remove() = Persistent.remove(this)

    /**
     * Template
     * ```json
     * {
     *      "uuid": $uuid,
     *      "x": $x,
     *      "y": $y,
     *      "task":
     *      {
     *          Task.json()
     *      }
     *      "icon":
     *      {
     *          BIcon.toRequest()
     *      }
     *      "label":
     *      {
     *          BLabel.json()
     *      }
     * }
     * ```
     */
    override fun toRequest(): JsonObject {
        val json = serialize()
        json.addProperty("profile", profile.toString())
        json.add("icon", back.toRequest())
        if (task !is ClientTask)
            json.remove("task")
        return json
    }

    /**
     * Template
     * ```json
     * {
     *      "uuid": $uuid,
     *      "x": $x,
     *      "y": $y,
     *      "task":
     *      {
     *          Task.json()
     *      }
     *      "icon":
     *      {
     *          BIcon.json()
     *      }
     *      "label":
     *      {
     *          BLabel.json()
     *      }
     * }
     * ```
     */
    override fun serialize(): JsonObject {
        val json = JsonObject()

        json.addProperty("uuid", uuid.toString())
        json.addProperty("x", posX)
        json.addProperty("y", posY)
        if (task != null)
            json.add("task", task!!.serialize())
        json.add("icon", back.serialize())
        json.add("label", front.serialize())

        return json
    }
}